### ISSUE TITLE
Dex Share follower doesn't receive data with new EU accounts having an email for username

### ISSUE THREAD

**Issue by psonnera** [NightscoutFoundation/xDrip#1882]
### Subject of the issue
New Dexcom users aren't proposed a username during account creation (only for linked accounts) and the username is the email address used to create the account. Dex Share follower doesn't receive any data with the email address in account credentials (checked correct server, username and password). Using a kid linked account with a username is functional.

### Your environment
I don't believe this is related to xDrip+ but Dex server recent changes.

### Expected behavior
Dex Share follower should receive data from Dex share servers.

### Actual behavior
No data is received by xDrip+ even if they are present in the vendor app followers.

### Your environment
I don't believe this is related to xDrip+ but Dex server recent changes.

### Expected behavior
Dex Share follower should receive data from Dex share servers.

### Actual behavior
No data.

### Steps to reproduce the behavior:
Create a new Dex account, log in the vendor master app, add a follower (either vendor app or xDrip+) and enable share. Accept the invitation in the vendor follower app to check functionality, set xDrip+ to Dex Share follower with matching credentials and server. And wait...

**Comment by tolot27** [comment]
Likely, Dexcom tries to enforce their API documented at https://developer.dexcom.com.

I still don't know the source of the xDrip implementation, because nothing is documented.

**Comment by psonnera** [comment]
@tolot27 I'm pretty sure it was documented there but even time travel machine doesn't have the original site... https://web.archive.org/web/20170919194707/https://developer.dexcom.com/ maybe @StephenBlackWasAlreadyTaken remembers or has some legacy documentation?

**Comment by psonnera** [comment]
@tolot27 @Navid200 which debug tags could be enabled to allow me to give more information on what's going on?

**Comment by fzelins** [comment]
I can confirm this bug. I have created new dexcom account, now my sensor expired so I deleted dexcom app, installed again and created new account. All new dexcom account have email as login. I have also completely reinstalled xDrip app and I am still getting error:

ShareFollowCallback
01:55 Get Share Data Failed: java.io.EOFException: End of input at line 1 column 1 path $

ShareFollowDL
01:55 Interceptor Body size: 0
01:55 Session id appears valid so going direct to get data
01:55 Setting message: 01: 55 Connectivity problem reaching Share servers

ShareFollowCallback
01:55 Get Share Data Failed: java.io.EOFException: End of input at line 1 column 1 path $

ShareFollowDL
01:55 Interceptor Body size: 0
01:55 Session id appears valid so going direct to get data

**Comment by Navid200** [comment]
@psonnera So sorry I missed your question.
The only tag I am aware of is this:
https://github.com/NightscoutFoundation/xDrip/wiki/Enabling-Debugging-for-OB1-collector

@tolot27 knows about this subject (Dexcom share) more than anyone else I know.

**Comment by samgray85** [comment]
> I can confirm this bug. I have created new dexcom account, now my sensor expired so I deleted dexcom app, installed again and created new account. All new dexcom accounts now have email as login. I have also completely reinstalled xDrip app and I am still getting error:
> 
> ShareFollowCallback 01:55 Get Share Data Failed: java.io.EOFException: End of input at line 1 column 1 path $
> 
> ShareFollowDL 01:55 Interceptor Body size: 0 01:55 Session id appears valid so going direct to get data 01:55 Setting message: 01: 55 Connectivity problem reaching Share servers
> 
> ShareFollowCallback 01:55 Get Share Data Failed: java.io.EOFException: End of input at line 1 column 1 path $
> 
> ShareFollowDL 01:55 Interceptor Body size: 0 01:55 Session id appears valid so going direct to get data

I have the same issue: https://github.com/Navid200/xDrip/discussions/146

**Comment by samgray85** [comment]
> I can confirm this bug. I have created new dexcom account, now my sensor expired so I deleted dexcom app, installed again and created new account. All new dexcom accounts now have email as login. I have also completely reinstalled xDrip app and I am still getting error:
> 
> ShareFollowCallback 01:55 Get Share Data Failed: java.io.EOFException: End of input at line 1 column 1 path $
> 
> ShareFollowDL 01:55 Interceptor Body size: 0 01:55 Session id appears valid so going direct to get data 01:55 Setting message: 01: 55 Connectivity problem reaching Share servers
> 
> ShareFollowCallback 01:55 Get Share Data Failed: java.io.EOFException: End of input at line 1 column 1 path $
> 
> ShareFollowDL 01:55 Interceptor Body size: 0 01:55 Session id appears valid so going direct to get data

@fzelins and @psonnera did you have any luck resolving

**Comment by tolot27** [comment]
I've fixed this issue and will create a PR soon. Still have to implement some tests for it.

**Comment by samgray85** [comment]
> I've fixed this issue and will create a PR soon. Still have to implement some tests for it.

Happy to test it for you if that helps

**Comment by KillaOfDawn** [comment]
I'm also affected by this and happy to test if required :)

**Comment by sudzwerg80** [comment]
Hello same problem here.would test it too.
Steffen

**Comment by sudzwerg80** [comment]
Hello Tolot27,
how can we help to solve this thread.
i want to use xdrip but i have to uninstall dexcom g6 app what is not so good because of my doc.
can i build an apk with your fix on my own ? when yes how ?

**Comment by usernamebeep** [comment]
following since it's still not working..

**Comment by Navid200** [comment]
@tolot27 I have added email format as one of the accepted options and opened a PR:
https://github.com/NightscoutFoundation/xDrip/pull/1974
I cannot test it because I don't have a Dex share account.

For all I know, you were going to release your fix  any minute now.  But, I have no idea.
I can close my PR if needed.

**Comment by KillaOfDawn** [comment]
Hey @Navid200 I have a feeling @tolot27's fix was a little more involved than adding the email format from looking at his commit here: https://github.com/tolot27/xDrip/commit/1a4079974a63040411c8d4ac9b02fd400baabb4a

I haven't been able to compile this version and run it myself but likelihood is that its my inexperience with Android apps.

Hope this helps :)

**Comment by Navid200** [comment]
>  I have a feeling @tolot27's fix was a little more involved than adding the email format from 

Yes, you may be correct.  
But, I don't know if you read my message.   

"I cannot test this" is much more condemning than "this may not be as involved".
So, I don't know how else to clarify that this is just an attempt, and may absolutely be useless.  It may just be forgotten, and I will be quite happy with that.

**Comment by tolot27** [comment]
> There are parents who cannot follow their children.

Maybe we can clarify this or you can help me to resolve a misunderstanding from my side. As far as I know, parents can still follow their children (with xDrip and Dexcom Follow) if they have created a Dexcom account with a nickname instead of an email address and using the Dexcom App or xDrip as collector. At least, I can follow my son.
At the time, Dexcom switched to email addresses as account names, only newly created user accounts are affected. Those new users/parents cannot switch from the Dexcom app to xDrip as collector and likely did not use xDrip before. If they had, they have likely used another share service like Nightscout and just discovered that xDrip also supports Dexcom Share. Those users/parents can still follow their children using these services.
To conclude and as it mentioned in the issues title, only new users are affected which like to use xDrip instead of Dexcom. Those users cannot switch yet.

Please comment only, if I'm wrong. Otherwise, give me a few more hours to work on an apk for testing.

**Comment by tolot27** [comment]
A debug apk for testing can be downloaded: [xDrip_20220131_b9d7a8027_dexPreAuth_debug.apk]( https://www.dropbox.com/s/6zq92llxghcl9hd/xDrip_20220131_b9d7a8027_dexPreAuth_debug.apk?dl=0). Please be aware that it cannot be installed over the existing app. Hence, make a backup of settings and database before you uninstall it.

**Comment by samgray85** [comment]
Ive download but it says cant open file. Samsung s10

**Comment by KillaOfDawn** [comment]
Hey @tolot27 thanks for that, I just downloaded and installed the app. I ran the set up wizard, selected the Dexcom Share option, entered the username and password I normally enter to the Dexcom G6 app and unselected the US option.

I received the following error:

![20220131_235934](https://user-images.githubusercontent.com/71264946/151892664-d97f6fdd-4944-438d-ac70-a73c4225527e.jpg)

I'm happy to help debug further if needed :)

**Comment by samgray85** [comment]
I did the same as @KillaOfDawn .

Unticked US (I'm  Australia), entered share username and password - confirmed these still work by using them to log into clarity.

I am a newish dexcom user with an email address for a username

![image](https://user-images.githubusercontent.com/86637332/151894867-19d51c9a-48a5-497a-8a5e-8feed8819e0b.png)

**Comment by tolot27** [comment]
@KillaOfDawn @samgray85 Please can you set the extra logging tag "ShareFollowDL:v", force xDrip to stop (settings -> apps) and start it again. Then please have a look at the logs. There should be some entries from ShareFollowDL like "Creating instance for ...". Please post the URL or a screenshot. I need the first log entries after startup.

**Comment by samgray85** [comment]
@tolot27 sorry you are going to have to walk me through how to set the extra logging tag

I have no programming experience.

**Comment by KillaOfDawn** [comment]
Hey @tolot27 please find a screenshot of the logs below:

![Screenshot_20220201-075919_xDrip+](https://user-images.githubusercontent.com/71264946/151932611-61c7a1db-23bf-4bf9-bd3f-d403670e080e.jpg)

I have some experience with adb but not with logcat. I should be able to pick up instructions though if required :)

**Comment by samgray85** [comment]
@tolot27  @KillaOfDawn 

Mine is working now. Thank you

![image](https://user-images.githubusercontent.com/86637332/151937257-fad9d41a-8cc0-4738-9f77-61386568529c.png)

![image](https://user-images.githubusercontent.com/86637332/151937266-1681c220-14d3-4bf7-abf2-b0d3a0cc1c77.png)

**Comment by KillaOfDawn** [comment]
@samgray85 Interesting, what changed that made it work suddenly?

**Comment by samgray85** [comment]
I just did what @tolot27 told me

Went:

- Settings
- Less common settings
- Extra logging settings
- Extra tags for logging
- added: ShareFollowDL:v

Then went into my phone settings > apps > xDrip+ > Force Stop

Reopened and it worked

**Comment by tolot27** [comment]
@KillaOfDawn Please can you verify your account data again? I suspect an error in either account name or password. Does it contain some special characters? In which country do you live? Maybe, your account is on the US server. Please can you try it?

> I should be able to pick up instructions though if required :)

If you still cannot get it to work, I need the logcat output.

**Comment by KillaOfDawn** [comment]
Hey @tolot27 Yep it was a problem on my end, I seemed to have set a password previously that was > 20 characters and this is likely what was causing it not to work. I reset the password and it all works now :)

Thank you so much for your help and thank you @samgray85 as well!

**Comment by sudzwerg80** [comment]
> A debug apk for testing can be downloaded: [xDrip_20220131_b9d7a8027_dexPreAuth_debug.apk](https://www.dropbox.com/s/6zq92llxghcl9hd/xDrip_20220131_b9d7a8027_dexPreAuth_debug.apk?dl=0). Please be aware that it cannot be installed over the existing app. Hence, make a backup of settings and database before you uninstall it.

it works on samsung s7
![Screenshot_20220201-104857_xDrip+](https://user-images.githubusercontent.com/51017703/151946664-08be238c-07d6-4f32-a8d6-4cf2272755a1.jpg)
 i am low sorry for that

**Comment by sudzwerg80** [comment]
thank you

**Comment by tolot27** [comment]
> I seemed to have set a password previously that was > 20 characters and this is likely what was causing it not to work. I reset the password and it all works now :)

I'll verify the password length. As far as I know, xDrip does not restrict its length but maybe Dexcom?

**Comment by KillaOfDawn** [comment]
> > I seemed to have set a password previously that was > 20 characters and this is likely what was causing it not to work. I reset the password and it all works now :)
> 
> I'll verify the password length. As far as I know, xDrip does not restrict its length but maybe Dexcom?

Yeah it seems to be on Dexcom's end, initially through their setup process (I assume through the G6 app) I was able to set a 22 length character but once I logged onto Clarity this morning I saw this in the password change field:

![image](https://user-images.githubusercontent.com/71264946/151951464-05e6adc7-9f16-4fa6-a019-587c20173950.png)

**Comment by cmarcosmoneo** [comment]
As additional info in case it can help, my username is not an email address but includes a dot in the middle.
I cannot use Xdrip+ since 30th November but I have tested this debug apk and it works fine.
So thank you so much for this and I am available of any further testing is needed.

**Comment by Navid200** [comment]
@tolot27 Thanks for fixing this
Adding it to the release notes shortly.

**Comment by alfiearthur** [comment]
Please, please, please can we incorporate this feature into the main Xdrip application? It's essential for European users and, at the moment, we have to switch off updates to prevent the feature being overwritten with the latest version. It's a life-saver for hypo-unaware Xdrip users outside the USA.

Thank you so much for developing it.

Alfie

**Comment by Navid200** [comment]
@alfiearthur This was already merged.  What is it that you want to be incorporated?
Can you be more specific plaease?

**Comment by alfiearthur** [comment]
Hi Navid,

Thank you for your quick reply. 

The problem is that when switching input to Hardware Data Source to ‘Dex Share Follower’ I enter my ‘Share Username’ and ‘Share Password’ and the next screen pops up to ask if I am based in the USA. I am in the UK so I untick ‘this location is in the USA’ and everything works. I’m not a programmer, but I think this directs Xdrip to the European Dexcom server which is https://uam2.dexcom.com, while the US server is https://uam <https://uam1.dexcom.com> 1.dexcom.com. 

This appears on the first attempt to change the hardware data source to ‘Dex Share Follower’ on the xDrip_20220131_b9d7a8027_dexPreAuth_debug.apk. On subsequent efforts to change the hardware data source to ‘Dex Share Follower’ it doesn’t appear again and I’m pretty sure it reverts to the US Dexcom server.

On the latest Xdrip release, ‘this location is in the USA’ pop with a tick box, doesn’t appear.

Something changed at 09:00 on Wednesday, November 30 2022 which meant XDrip couldn’t connect to Dexcom Follower any more. This simultaneously happened with Tidepool, which is no longer connecting to our Dexcom Follow account. My guess is that Dexcom were running our UK account on the US server and they transferred our account to the European server at that time.

Our T1D son has no hypoawareness and doesn’t wake up to alarms so I spend the night with him. With the xDrip_20220131_b9d7a8027_dexPreAuth_debug.apk version, I have now been able to reconnect to Dexcom Share and we’re back in our own beds. I let Dexcom update the version and we were back to square 1, where Xdrip couldn’t find the European Dexcom server.

 

In summary, I don’t think the option to choose tick/untick ‘I am in the USA’ when changing the Hardware Data Source to ‘Dex Share Follower’ is working in the current version and, even in the beta version it only appeared on the first attempt.

 

I hope this makes sense! I’ve disabled updates to preserve the functionality for now.

 

I totally realise it’s a massive faff to test this stuff and I am deeply grateful (we all are) for all the work that you guys put in to help our children stay alive.

 

Thank you for everything,

 

Rob

 

Rob Arthur

  _____  

Enderley House, Amberley, Gloucestershire GL5 5AW 
Tel: +44 (0) 1453 872121   Mob: +44 (0) 7581 242515   Skype: robarthur2

  _____

**Comment by jamorham** [comment]
Hi Rob,

Can you try with the latest released xDrip nightly? (See the releases section here)

The setting for which server to use can be found in the cloud upload section of xDrip main settings. The same server is used for upload/download. If you change settings it is also possible you need to go to xDrip status and restart the collector to start using the new settings.

Can you also describe your set up in terms of what is the configuration for your son's phone? Is it Android and if so which app are you running there? Which android version is your phone running as well?

**Comment by alfiearthur** [comment]
Hi Jam or Ham (tough choice, but ham, please, if it’s smoked)

Alfie is at school at the moment. He has a Pixel 4a. I’ll try downloading the xDrip nightly today and let you know.

 

Many thanks,

 

Rob

 

Rob Arthur

  _____  

Enderley House, Amberley, Gloucestershire GL5 5AW 
Tel: +44 (0) 1453 872121   Mob: +44 (0) 7581 242515   Skype: robarthur2

  _____

**Comment by jamorham** [comment]
@alfiearthur I think if you've been having problems then the best way to resolve this is to install the latest nightly on both your son's phone and your own. If you are using the official manufacturer app on your son's phone then you can continue to do this and simply set up xDrip there in companion mode and it will run side by side. You can then make it an xDrip sync master and set your follower phone to be a follower and then you don't have to worry about any issues using dex share following. If you need any help with these steps I would suggest creating a discussion here instead of appending to this issue and describe what you need help with. The only real use case that I know of for dex share is if your follower/master is on another platform like apple.

**Comment by alfiearthur** [comment]
That’s great advice. 

The only reason we went to Dex Share is the demise of the free Mongo database.

We have been using Tidepool to analyse the data and share it with our clinic, but that has also broken following the transfer from Dexcom US server to their EU server.

 

Many thanks,

 

Rob

 

Rob Arthur

  _____  

Enderley House, Amberley, Gloucestershire GL5 5AW 
Tel: +44 (0) 1453 872121   Mob: +44 (0) 7581 242515   Skype: robarthur2

  _____

**Comment by jamorham** [comment]
@alfiearthur xDrip has a native tidepool uploader so you can use that if you like.

**Comment by alfiearthur** [comment]
Thank you. I’ll have a play with that right now,

 

Rob

 

Rob Arthur

  _____  

Enderley House, Amberley, Gloucestershire GL5 5AW 
Tel: +44 (0) 1453 872121   Mob: +44 (0) 7581 242515   Skype: robarthur2

  _____

**Comment by alfiearthur** [comment]
The l;atest nightly version does indeed include an option for non-USA servers. I found it [here](https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.08.03). Everything is now working fine.
Just trying to get Tidepool to catch up with you.

Many, many thanks for all your help,

Rob (and Alfie!)
